File: Top_100_Liked_Questions/46._Permutations/solution.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        # swap method
        ans = []
        
        def backtrack(first=0):
            if first == len(nums):
                ans.append(nums[:])
                return
            
            for i in range(first, len(nums)):
                nums[first], nums[i] = nums[i], nums[first]
                backtrack(first+1)
                nums[first], nums[i] = nums[i], nums[first]
        
        backtrack()
        
        return ans
    
    
    def permute2(self, nums: List[int]) -> List[List[int]]:
        # decreasing pool method
        ans = []

        def helper(remainingPool, currList=[]):
            if not remainingPool:
                ans.append(currList)
                return
            
            for i,v in enumerate(remainingPool):
                copyRemainingPool = remainingPool[:]
                del copyRemainingPool[i]
                helper(copyRemainingPool, currList + [v])
        
        helper(nums)
        
        return ans

File: Top_100_Liked_Questions/46._Permutations/solution.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    let ans = [];
    
    let backtrack = function(first=0) {
        if (first === nums.length) {
            ans.push([...nums]);
            return;
        }
        
        for (let i=first; i<nums.length; i++) {
            [nums[first], nums[i]] = [nums[i], nums[first]];
            backtrack(first+1);
            [nums[first], nums[i]] = [nums[i], nums[first]];
        }
    }
    
    backtrack();
    
    return ans;
}

var permute2 = function(nums) {
    let ans = [];

    let helper = function(rem, per=[]) {
        if (rem.length === 0) {
            ans.push(per);
            return;
        }
        
        for (let i=0; i<rem.length; i++) {
            remCopy = [...rem];
            remCopy.splice(i,1);
            helper(remCopy, [...per, rem[i]]);
        }
    }
    
    helper(nums);
    
    return ans;
};